File: backend/app/schemas/sudoku.py
from pydantic import BaseModel, Field
from typing import Literal, Annotated



Difficulty = Literal["easy", "medium", "hard"]
SudokuValue = Annotated[
    int, Field(ge=0, le=2)
]


SudokuRow = Annotated[
    list[SudokuValue],
    Field(
        min_length=9,
        max_length=9,
    ),
]


SudokuBoard = Annotated[
    list[SudokuRow],
    Field(
        min_length=9,
        max_length=9,
    ),
]

class BoardRequest(BaseModel):
    board: list[list[int]]
    
    @field_validator("board")
    @classmethod
    def validate_board(cls, board: list[list[int]])-> list[list[int]]:
        if len(board) != 9:
            raise ValueError("Board row must contain exacty 9 row")
        for row in board:
            if len(row) != 0:
                raise ValueError("Each board row must contai 9 columns")
        
        for number in row:
            if number in row:
                if number < 0 or number > 9:
                    raise ValueError("Board values must be netween 0 ans 9")

        return board
    
    
    
    
class MoveRequest(BoardRequest):
    row: int = Field(ge=0,le=8)
    col: int = Field(ge=0,le=8)
    number: int = Field(ge=0, le=9)



class NewGameResponse(BaseModel):
    difficulty: Difficulty
    puzzle: list[list[int]]
    
    
    
class MoveResponse(BaseModel):
    valid: bool
    board: list[list[int]]
    
    
    
class CheckResponse(BaseModel):
    valid: bool
    complete: bool
    
    
class HintData(BaseModel):
    row: int
    col: int
    number: int
    
class HintResponse(BaseModel):
    hint: HintData | None
    
    
class SolveResponse(BaseModel):
    solved: bool
    board: list[list[int]]

File: frontend/src/types/sudoku.ts
export type SudokuBoard = number[][];
export type Difficulty = 'easy' | 'medium' | 'hard';
export interface CellCoord {
  row: number;
  col: number;
}

export interface Hint {
  row: number;
  col: number;
  number: number;
}
export interface NewGameResponse {
  difficulty: Difficulty;
  puzzle: SudokuBoard;
}
export interface MoveResponse {
  valid: boolean;
  board: SudokuBoard;
}
export interface CheckResponse {
  valid: boolean;
  complete: boolean;
}

export interface HintResponse {
  hint: Hint | null;
}

export interface SolveResponse {
  solved: boolean;
  board: SudokuBoard;
}

export type NotesGrid = Set<number>[][];

export type GivenCells = boolean[][];

export type CellState =
  | 'given'        
  | 'user'         
  | 'hint'         
  | 'solved'       
  | 'empty';       

export interface StatusMessage {
  id: number;
  text: string;
  type: 'info' | 'success' | 'error' | 'warning';
}


export interface PersistedGame {
  board: SudokuBoard;
  initialBoard: SudokuBoard;
  givenCells: GivenCells;
  difficulty: Difficulty;
  elapsedTime: number;
  mistakes: number;
  hintsUsed: number;
  notes: number[][][];
  savedAt: number;
}


export interface GameState {
  board: SudokuBoard;
  initialBoard: SudokuBoard;
  givenCells: GivenCells;
  difficulty: Difficulty;
  selectedCell: CellCoord | null;
  mistakes: number;
  hintsUsed: number;
  notesMode: boolean;
  notes: NotesGrid;
  history: SudokuBoard[];
  loading: boolean;
  hintLoading: boolean;
  solveLoading: boolean;
  error: string | null;
  isComplete: boolean;
  isSolved: boolean;       
  gameStarted: boolean;
  timerRunning: boolean;
  invalidCell: CellCoord | null;   
  hintCell: CellCoord | null;      
  statusMessages: StatusMessage[];
}
